### ISSUE TITLE
Logout using multipe identityservers

### ISSUE THREAD

**Issue by devlz303** [IdentityServer/Thinktecture.IdentityServer3#870]
I have a situation in which I want to use multiple identityservers "connected" to each other and now I want to log out of the web app and be able to return to the web app after logging out.
In the sample applications this is "simple", because there is only one IdentityServer involved. After logging out the logged out page is shown, with a link back to the web application (according to the PostLogoutRedirectUri of the client).

In my situation it is as follows. I have 3 applications:
1. A web app that needs id tokens (and access tokens for web api's).
2. A security token service for the web app with a custom user service to "manage" app specific users and roles.
3. An global identity provider used for the actual authentication of the users (SSO).

The security token service (STS) and global identity provider (SSO) both use Thinktecture IdentityServer3. The STS does not allow local login, but uses SSO instead (STS is a client of SSO).
The SSO service uses only local login (for now I use InMemoryUsers).

When a user makes a request to a secured page, the user is redirected to the STS which in turn redirects to the external identity provider. The user signs in and is redirected back to the STS.
The STS looks up the user in it's own app specific database and then creates an claimsidentity with specific roles for the application. Lastly the user is being redirected to the web app.
This all seems to work as expected.

But now I ran into a problem when signin out. When I don't add the login hint I first end up at the logout confirmation page of the STS, asking if I would like to sign out.
When I click Yes to sign out, the sign out at the STS happens and then I get redirected to the endsession endpoint of the external identity provider (SSO). This is because in the AuthenticationController of IdSrv3 a call is made to the SignOutOfExternalIdP() causing the redirect.
I now get signed out of the external identityprovider and the logged out page of SSO is shown with only the message: You are now logged out

I know it's because of the missing PostLogoutRedirectUri and IdTokenHint it is not possible to show the redirect, but I don't know how to solve this.

Is there anybody out there having the same issue or know a solution for my situation?

Thanks!

**Comment by devlz303** [comment]
Well, I've managed to find one peace of the puzzle. After configuring logging I ran into the following message when signing out of the external identity provider:

Thinktecture.IdentityServer.Core.Validation.EndSessionRequestValidator: User is anonymous. Aborting logout.

Somehow the session was already killed. Using Fiddler I was able to find out why.
Because all the applications in the development environment are running on localhost (different ports) there is a problem with the authentication cookies that have the same name on the different apps.
The cookies idsrv and idsrv.session where set by the external identity provider and then overwritten by the security token service use by the web app.
Setting the AuthenticationOptions.CookieOptions.Prefix for one of the identityservers fixed this problem.

**Comment by devlz303** [comment]
I have got it working now. After solving the cookie problem, the only thing left was storing the id token for the external identity provider and set it in the ProtocolMessage.IdTokenHint when making the LogoutRequest.
I still have to refactor and clean up the code now, because I made a bit of a mess of it.
To be able to redirect to the actual web application I have set the PostLogoutRedirectUri to the home page of the web application instead of the STS so after signing out the use can click on the link to get back to the web application.
